File: 329.longest_increasing_path_in_a_matrix/longestIncreatingPath.py
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        self.n = len(matrix)
        if self.n == 0:
            return 0
        self.m = len(matrix[0])
        self.dp = [[0]*self.m for i in range(self.n)]
        self.flag = [[0]*self.m for i in range(self.n)]
        ans = 0

        for i in range(self.n):
            for j in range(self.m):
                self.dp[i][j] = self.search(i, j, matrix)
                ans = max(ans, self.dp[i][j])

        return ans

    def search(self,x, y, matrix):
        dx = [1,-1,0,0]
        dy = [0,0,1,-1]
        flag = self.flag
        dp = self.dp
        if(flag[x][y] != 0):
            return dp[x][y]
        ans = 1
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx and nx < self.n and 0 <= ny and ny < self.m:
                if matrix[x][y] > matrix[nx][ny]:
                    ans = max(ans, self.search(nx, ny, matrix)+1)

        flag[x][y] = 1
        dp[x][y] = ans
        return ans

File: 329.longest_increasing_path_in_a_matrix/longestIncreatingPath.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var dx = [1,-1,0,0];
var dy = [0,0,1,-1];//the step

var longestIncreasingPath = function(matrix) {
    "use strict";
    var search = function (x,y,matrix) {
        if(flag[x][y] !== 0){
            return dp[x][y]; //has been searched
        }
        var ans = 1;
        var nx,ny;
        for(var i = 0; i < 4; i++){
            nx = x + dx[i];
            ny = y + dy[i];
            if(0 <= nx && nx < n && 0 <= ny && ny < m){
                console.log(matrix[x][y],matrix[nx][ny]);
                if(matrix[x][y] > matrix[nx][ny]){

                    ans = Math.max(ans, search(nx, ny, matrix)+1);
                    console.log(ans);
                }
            }
        }
        flag[x][y] =1;
        dp[x][y]=ans;
        return ans;
    };

    var n = matrix.length;
    if(n == 0){
        return 0;
    }
    var m = matrix[0].length;
    // console.log(n,m);
    var dp = new Array(n);
    var flag = new Array(n);
    for(var i = 0; i < n; i++){
        dp[i] = new Array(m);
        dp[i].fill(0);
        flag[i] = new Array(n);
        flag[i].fill(0);
    }//initilize the 2d array
    var ans = 0;
    for(let i = 0; i < n; i++){
        for(let j = 0; j < m; j++){
            dp[i][j] = search(i,j,matrix);
            ans = Math.max(dp[i][j],ans);
            // console.log(dp[i][j]);
        }
    }
    return ans;

};


// console.log(longestIncreasingPath([[9,9,4],[6,6,8],[2,1,1]]))
console.log(longestIncreasingPath([[1,2,3,4]]));